watertap/unit_models/crystallizer.py

```diff
@@ -0,0 +1,696 @@
+            units=pyunits.dimensionless,
+            doc="Crystallizer solids yield"
+            )
+
+        self.product_volumetric_solids_fraction = Var(
+            initialize=0.25,
+            bounds=(1e-8, 1),
+            units=pyunits.dimensionless,
+            doc = "Volumetric fraction of solids in slurry product (i.e. solids-liquid mixture)."
+            )
+
+        self.temperature_operating = Var(
+            initialize=298.15,
+            bounds=(273, 1000),
+            units=pyunits.K,
+            doc="Crystallizer operating temperature: boiling point of the solution."
+            )
+
+        self.pressure_operating = Var(
+            initialize=1e3,
+            bounds=(0.001, 1e6),
+            units=pyunits.Pa,
+            doc="Operating pressure of the crystallizer."
+            )
+
+        self.magma_density = Var(
+            initialize=250,
+            bounds=(1, 5000),
+            units=pyunits.kg/pyunits.m**3,
+            doc="Magma density, i.e. mass of crystals per unit volume of suspension"
+            )
+
+        self.slurry_density = Var(
+            initialize=1000,
+            bounds=(1, 5000),
+            units=pyunits.kg/pyunits.m**3,
+            doc="Suspension density, i.e. density of solid-liquid mixture before separation"
+            )
+
+        self.work_mechanical = Var(
+            self.flowsheet().config.time,
+            initialize=1e5,
+            bounds=(-5e6, 5e6),
+            units=pyunits.kJ/pyunits.s,
+            doc="Crystallizer thermal energy requirement"
+            )
+
+        self.D_crystallizer = Var(
+            initialize=3,
+            bounds=(0, 25),
+            units=pyunits.m,
+            doc="Minimum diameter of crystallizer"
+            )
+
+
+        self.h_slurry = Var(
+            initialize=3,
+            bounds=(0, 25),
+            units=pyunits.m,
+            doc="Slurry height in crystallizer"
+            )
+
+        self.H_crystallizer = Var(
+            initialize=3,
+            bounds=(0, 25),
+            units=pyunits.m,
+            doc="Crystallizer height"
+            )
+
+        self.magma_circulation_flow_vol = Var(
+            initialize=1,
+            bounds=(0, 100),
+            units=pyunits.m**3/pyunits.s,
+            doc="Minimum circulation flow rate through crystallizer heat exchanger"
+            )
+
+        self.relative_supersaturation = Var(
+            solute_set,
+            initialize=0.1,
+            bounds=(0, 100),
+            units=pyunits.dimensionless
+        )
+
+        self.t_res =  Var(
+            initialize=1,
+            bounds=(0, 10),
+            units=pyunits.hr,
+            doc="Residence time in crystallizer"
+            )
+
+        self.V_suspension = Var(
+            initialize=0,
+            bounds=(0, None),
+            units=pyunits.m**3,
+            doc="Crystallizer minimum active volume, i.e. volume of liquid-solid suspension"
+            )
```

Comment:
I'd also write this out as `volume`, is this also not the crystallizer volume? Do you have a variable for the crystallizer volume?

Reply:
This is not the crystallizer volume. It is the volume of the solid/liquid mixture in the crystallizer. I have no variable for the crystallizer volume at the moment.